environment/src/main/java/com/sequenceiq/environment/environment/flow/modify/proxy/action/AbstractEnvProxyModificationAction.java

```diff
@@ -34,31 +36,25 @@ protected AbstractEnvProxyModificationAction(Class<P> clazz) {
     protected EnvProxyModificationContext createFlowContext(FlowParameters flowParameters,
             StateContext<EnvProxyModificationState, EnvProxyModificationStateSelectors> stateContext, P payload) {
         Map<Object, Object> variables = stateContext.getExtendedState().getVariables();
-        Optional<ProxyConfig> previousProxyConfig =
-                (Optional<ProxyConfig>) variables.computeIfAbsent(PREVIOUS_PROXY_CONFIG, k -> getPreviousProxyConfig(payload));
-        return new EnvProxyModificationContext(flowParameters, previousProxyConfig.orElse(null));
+        ProxyConfig proxyConfig = (ProxyConfig) variables.computeIfAbsent(PROXY_CONFIG, k -> getProxyConfig(payload));
+        ProxyConfig previousProxyConfig = (ProxyConfig) variables.computeIfAbsent(PREVIOUS_PROXY_CONFIG, k -> getPreviousProxyConfig(payload));
+        return new EnvProxyModificationContext(flowParameters, proxyConfig, previousProxyConfig);
+    }
+
+    private ProxyConfig getProxyConfig(P payload) {
+        return proxyConfigService.getByCrn(payload.getProxyConfigCrn());
     }
 
-    private Optional<ProxyConfig> getPreviousProxyConfig(P payload) {
-        return Optional.ofNullable(environmentService.findEnvironmentByIdOrThrow(payload.getResourceId()).getProxyConfig());
+    private ProxyConfig getPreviousProxyConfig(P payload) {
+        return proxyConfigService.getByEnvironmentCrn(payload.getResourceCrn());
```

Comment:
Does it mean the current association between the environment and proxy config, before the modification?

Reply:
Exactly. But thanks for pointing it out as it may throw a NotFoundException, this has to be handled for the use case of adding a proxy to the cluster.